doc/source/reference/c-api/data_memory.rst

```diff
@@ -0,0 +1,96 @@
+Memory management
+-----------------
+
+The `numpy.ndarray` is a python class. It requires additinal memory allocations
+to hold `numpy.ndarray.strides`, `numpy.ndarray.shape` and
+``numpy.ndarray.data`` attributes. These attributes are specially allocated
+after creating the python object in `__new__`. The ``strides`` and
+``dimensions`` are stored in a piece of memory allocated internally.
+
+These allocations are small relative to the ``data``, the homogeneous chunk of
+memory used to store the actual array values (which could be pointers in the
+case of ``object`` arrays). Users may wish to override the internal data
+memory routines with ones of their own. They can do this by using
+``PyDataMem_SetHandler``, which uses a ``PyDataMem_Handler`` structure to hold
+pointers to functions used to manage the data memory. The calls are wrapped
+by internal routines to call :c:func:`PyTraceMalloc_Track`,
+:c:func:`PyTraceMalloc_Untrack`, and will use the `PyDataMem_EventHookFunc`
+mechanism. Since the functions may change during the lifetime of the process,
+each `ndarray` carries with it the functions used at the time of its
+instantiation, and these will be used to reallocate or free the data memory of
+the instance.
+
+.. c:type:: PyDataMem_Handler
+
+    A struct to hold function pointers used to manipulate memory
+
+    .. code-block:: c
+
+        typedef struct {
```

Comment:
It would be good to reference the builtin `PyMemAllocatorEx` somewhere in this section and compare the two (https://docs.python.org/3/c-api/memory.html#customize-memory-allocators).

Reply:
Will add. The original `npy_*_cache` set of functions uses `PyDataMem_*` functions, which it seems preceeded these more sophisticated interfaces and directly used `malloc`/`calloc`/`free`.

Reply:
Documented.  The Python ones do not use our non-documented-but-public `PyDataMem_EventHookFunc` callbacks. So I could see a path where we deprecate `PyDataMem_EventHookFunc` and move to the Python memory management strategies, although that would mean
- if someone by chance implemented a `PyDataMem_EventHookFunc` callback it would no longer work.
- in order to override data allocations, a user would also override other random memory allocations